amigo-lib/src/main/java/me/ele/amigo/release/ApkReleaseActivity.java

```diff
@@ -79,29 +85,36 @@ protected void onCreate(Bundle savedInstanceState) {
             @Override
             public void run() {
                 super.run();
-
                 DexReleaser.releaseDexes(demoAPk.getAbsolutePath(), dexDir.getAbsolutePath());
                 NativeLibraryHelperCompat.copyNativeBinaries(demoAPk, nativeLibraryDir);
                 dexOptimization();
-
-                handler.sendEmptyMessage(WHAT_DEX_OPT_DONE);
             }
         }.start();
     }
 
     private void dexOptimization() {
         File[] listFiles = dexDir.listFiles();
 
-        List<File> validDexes = new ArrayList<>();
+        final List<File> validDexes = new ArrayList<>();
         for (File listFile : listFiles) {
             if (listFile.getName().endsWith(".dex")) {
                 validDexes.add(listFile);
             }
         }
 
-        for (File dex : validDexes) {
-            new DexClassLoader(dex.getAbsolutePath(), optimizedDir.getAbsolutePath(), null, DexUtils.getPathClassLoader());
-            Log.e(TAG, "dexOptimization finished-->" + dex);
+        for (final File dex : validDexes) {
+            service.submit(new Runnable() {
+                @Override
+                public void run() {
+                    long startTime = System.currentTimeMillis();
+                    new DexClassLoader(dex.getAbsolutePath(), optimizedDir.getAbsolutePath(), null, DexUtils.getPathClassLoader());
+                    Log.e(TAG, String.format("dex %s consume %d ms", dex.getAbsolutePath(), System.currentTimeMillis() - startTime));
+                    int num = counter.incrementAndGet();
+                    if (num == validDexes.size()) {
+                        handler.sendEmptyMessage(WHAT_DEX_OPT_DONE);
+                    }
```

Comment:
Initiates an orderly shutdown in which previously submitted
 tasks are executed, but no new tasks will be accepted.